backend-comparison/src/persistence/base.rs

```diff
@@ -1,100 +1,105 @@
+use uuid::Uuid;
 
 #[derive(Default)]
-pub struct Persistence {
-    results: BenchmarkCommitResults,
+pub struct BenchmarkRecord {
+    backend: String,
+    device: String,
+    results: BenchmarkResult,
 }
 
-impl Persistence {
-    /// Updates the cached backend comparison json file with new benchmarks results.
-    ///
-    /// The file has the following structure:
-    ///
-    ///  {
-    ///    "GIT_COMMIT_HASH":
-    ///      {
-    ///        "BENCHMARK_NAME (OP + SHAPE)": {
-    ///          "BACKEND_NAME-DEVICE": {
-    ///            "TIMESTAMP": \[
-    ///              DURATIONS
-    ///           \]
-    ///         }
-    ///       }
-    ///    }
-    ///  }
-    pub fn persist<B: Backend>(benches: Vec<BenchmarkResult>, device: &B::Device) {
-        for bench in benches.iter() {
-            println!("{}", bench);
-        }
-        let cache_file = dirs::home_dir()
-            .expect("Could not get home directory")
-            .join(".cache")
-            .join("backend-comparison")
-            .join("db.json");
+/// Save the benchmarks results on disk.
+///
+/// The structure is flat so that it can be easily queried from a database
+/// like MongoDB.
+///
+/// ```txt
+///  [
+///    {
+///      "backend": "backend name",
+///      "device": "device name",
+///      "git_hash": "hash",
+///      "name": "benchmark name",
+///      "operation": "operation name",
+///      "shapes": ["shape dimension", "shape dimension", ...],
+///      "timestamp": "timestamp",
+///      "numSamples": "number of samples",
+///      "min": "duration in seconds",
+///      "max": "duration in seconds",
+///      "median": "duration in seconds",
+///      "mean": "duration in seconds",
+///      "variance": "duration in seconds"
+///      "rawDurations": ["duration 1", "duration 2", ...],
+///    },
+///    { ... }
+/// ]
+/// ```
+pub fn save<B: Backend>(
+    name: &str,
+    benches: Vec<BenchmarkResult>,
+    device: &B::Device,
+) -> Result<Vec<BenchmarkRecord>, std::io::Error> {
+    let cache_dir = dirs::home_dir()
+        .expect("Could not get home directory")
+        .join(".cache")
+        .join("backend-comparison");
 
-        let mut cache = Self::load(&cache_file);
-        cache.update::<B>(device, benches);
-        cache.save(&cache_file);
-        println!("Persisting to {:?}", cache_file);
+    if !cache_dir.exists() {
+        fs::create_dir_all(&cache_dir)?;
     }
 
-    /// Load the cache from disk.
-    fn load(path: &PathBuf) -> Self {
-        let results = match File::open(path) {
-            Ok(file) => serde_json::from_reader(file)
-                .expect("Should have parsed to BenchmarkCommitResults struct"),
-            Err(_) => HashMap::default(),
-        };
+    let uuid = Uuid::new_v4().simple().to_string();
+    let short_uuid = &uuid[..8];
+    let file_name = format!("benchmarks_{}_{}.json", name, short_uuid);
+    let file_path = cache_dir.join(file_name);
```

Comment:
I think using a timestamp is enough and will be better sorted on disk.

Reply:
I can add it back, my reasoning was that we can already sort by creation date but that's not as simple indeed.

Reply:
Oh and we can theoretically overwrite a file as I wrote initially.

Also it's handy to have the name  of the benchmark.

Reply:
yup keep the name, juste replacing the uuid with a timestamp, users should never run multiple benchmarks at the same time, the benchmarks won't be valid in this case anyways.